src/Gui/ImageView.cpp

```diff
@@ -0,0 +1,369 @@
+ *   write to the Free Software Foundation, Inc., 59 Temple Place,         *
+ *   Suite 330, Boston, MA  02111-1307, USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "PreCompiled.h"
+#ifndef _PreComp_
+# include <QAction>
+# include <QApplication>
+# include <QContextMenuEvent>
+# include <QClipboard>
+# include <QCursor>
+# include <QFileInfo>
+# include <QImageReader>
+# include <QLabel>
+# include <QMenu>
+# include <QMessageBox>
+# include <QMimeData>
+# include <QPainter>
+# include <QPixmap>
+# include <QPrintDialog>
+# include <QPrinter>
+# include <QScrollArea>
+# include <QScrollBar>
+#endif
+
+#include "ImageView.h"
+#include "BitmapFactory.h"
+
+using namespace Gui;
+
+ImageView::ImageView(QWidget* parent)
+    : MDIView(nullptr, parent)
+    , imageLabel(new QLabel)
+    , scrollArea(new QScrollArea)
+    , scaleFactor{1.0}
+    , dragging{false}
+{
+    imageLabel->setBackgroundRole(QPalette::Base);
+    imageLabel->setSizePolicy(QSizePolicy::Ignored, QSizePolicy::Ignored);
+    imageLabel->setScaledContents(true);
+
+    scrollArea->setBackgroundRole(QPalette::Dark);
+    scrollArea->setWidget(imageLabel);
+    scrollArea->setVisible(false);
+    setCentralWidget(scrollArea);
+    setAcceptDrops(true);
+    setWindowIcon(Gui::BitmapFactory().pixmap("colors"));
+}
+
+bool ImageView::loadFile(const QString& fileName)
+{
+    QImageReader reader(fileName);
+    reader.setAutoTransform(true);
+    QImage image = reader.read();
+    if (image.isNull()) {
+        QMessageBox::information(this, tr("Failed to load image file"),
+                                 tr("Cannot load file %1: %2")
+                                 .arg(fileName, reader.errorString()));
+        return false;
+    }
+
+    setImage(image);
+    setWindowFilePath(fileName);
+
+    return true;
+}
+
+void ImageView::setImage(const QImage& image)
+{
+    rawImage = image;
+    imageLabel->setPixmap(QPixmap::fromImage(image));
+    imageLabel->adjustSize();
+    scrollArea->setVisible(true);
+    scaleFactor = 1.0;
+}
+
+void ImageView::scaleImage(double factor)
+{
+    scaleFactor *= factor;
+#if QT_VERSION >= QT_VERSION_CHECK(5,15,0)
+    imageLabel->resize(scaleFactor * imageLabel->pixmap(Qt::ReturnByValue).size());
+#else
+    imageLabel->resize(scaleFactor * imageLabel->pixmap()->size());
+#endif
+
+    adjustScrollBar(scrollArea->horizontalScrollBar(), factor);
+    adjustScrollBar(scrollArea->verticalScrollBar(), factor);
+}
+
+void ImageView::adjustScrollBar(QScrollBar *scrollBar, double factor)
+{
+    scrollBar->setValue(int(factor * scrollBar->value()
+                            + ((factor - 1) * scrollBar->pageStep()/2)));
+}
+
+bool ImageView::canZoomIn() const
+{
+    return !isFitToWindow() && imageLabel->width() < 10000;
```

Comment:
Magic number -- there are several around here (200 in canZoomOut, etc.). My clang-tidy complains about them, suggesting using named constants.

Reply:
Added an identifier